File: library/src/main/java/com/github/javiersantos/piracychecker/SaltUtils.java
package com.github.javiersantos.piracychecker;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import java.util.Random;

/**
 * Credits to Aidan Follestad (afollestad)
 */
class SaltUtils {
    private static final String KEY_SALT = "piracy-salt";
    private static byte[] mSalt;

    private static void generateSalt(Context context) {
        mSalt = new byte[20];
        final Random randomGenerator = new Random();
        for (int i = 0; i < 20; ++i) {
            mSalt[i] = (byte) (randomGenerator.nextInt(600) - 300);
        }
        final String saltStr = getSaltString();
        PreferenceManager.getDefaultSharedPreferences(context)
                .edit().putString(KEY_SALT, saltStr).apply();
    }

    private static String getSaltString() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < mSalt.length; i++) {
            if (i > 0) {
                sb.append(" ");
            }
            sb.append(mSalt[i]);
        }
        return sb.toString();
    }

    private static byte[] bytesFromString(String string) {
        final String[] split = string.split(" ");
        final byte[] data = new byte[split.length];
        for (int i = 0; i < split.length; i++) {
            data[i] = Byte.parseByte(split[i]);
        }
        return data;
    }

    @SuppressWarnings("ConstantConditions")
    static byte[] getSalt(Context context) {
        if (mSalt == null) {
            final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
            if (prefs.contains(KEY_SALT)) {
                mSalt = bytesFromString(prefs.getString(KEY_SALT, null));
            }
            if (mSalt == null) {
                generateSalt(context);
            }
        }
        return mSalt;
    }
}

File: library/src/main/java/com/github/javiersantos/piracychecker/utils/SaltUtils.kt
package com.github.javiersantos.piracychecker.utils

import android.content.Context
import androidx.preference.PreferenceManager
import java.util.Random

/**
 * Credits to Aidan Follestad (afollestad)
 */
internal object SaltUtils {
    private const val KEY_SALT = "salty-salt"
    private var mSalt: ByteArray? = null
    
    private val saltString: String
        get() {
            val sb = StringBuilder()
            mSalt?.let {
                for (i in it.indices) {
                    if (i > 0) sb.append(" ")
                    sb.append(it[i].toString())
                }
            }
            return sb.toString()
        }
    
    private fun generateSalt(context: Context?) {
        mSalt = ByteArray(20)
        val randomGenerator = Random()
        mSalt?.let {
            for (i in 0..19) {
                it[i] = (randomGenerator.nextInt(600) - 300).toByte()
            }
        }
        context ?: return
        PreferenceManager.getDefaultSharedPreferences(context)
            .edit().putString(KEY_SALT, saltString).apply()
    }
    
    private fun bytesFromString(string: String): ByteArray {
        val split = string.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val data = ByteArray(split.size)
        for (i in split.indices) {
            data[i] = java.lang.Byte.parseByte(split[i])
        }
        return data
    }
    
    fun getSalt(context: Context?): ByteArray? {
        if (mSalt == null) {
            mSalt = context?.let {
                try {
                    val prefs = PreferenceManager.getDefaultSharedPreferences(context)
                    if (prefs.contains(KEY_SALT)) {
                        val saltFromPrefs = prefs.getString(KEY_SALT, null)
                        saltFromPrefs?.let { bytesFromString(it) }
                    } else null
                } catch (e: Exception) {
                    null
                }
            }
            if (mSalt == null) generateSalt(context)
        }
        return mSalt
    }
}